common/src/main/java/io/netty/util/internal/PriorityQueue.java

```diff
@@ -0,0 +1,259 @@
+/*
+ * Copyright 2015 The Netty Project
+ *
+ * The Netty Project licenses this file to you under the Apache License,
+ * version 2.0 (the "License"); you may not use this file except in compliance
+ * with the License. You may obtain a copy of the License at:
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ */
+package io.netty.util.internal;
+
+import java.util.AbstractQueue;
+import java.util.Arrays;
+import java.util.Iterator;
+import java.util.NoSuchElementException;
+import java.util.Queue;
+
+import static io.netty.util.internal.ObjectUtil.checkNotNull;
+import static io.netty.util.internal.PriorityQueueNode.INDEX_NOT_IN_QUEUE;
+
+/**
+ * A priority queue which uses natural ordering of elements. Elements are also required to be of type
+ * {@link PriorityQueueNode} for the purpose of maintaining the index in the priority queue.
+ * @param <T> The object that is maintained in the queue.
+ */
+public final class PriorityQueue<T extends PriorityQueueNode<T>> extends AbstractQueue<T> implements Queue<T> {
+    private Object[] queue;
```

Comment:
@ejona86 - Just to clarify are you suggesting taking the `Class<T>` as an argument to the constructor? Or do you have another way to do this?